src/pokemon-service/interactor/pokemon.go

```diff
@@ -0,0 +1,98 @@
+package interactor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/GerardoHP/ondemand-go-bootcamp/dto"
+	"github.com/GerardoHP/ondemand-go-bootcamp/entity"
+	"github.com/GerardoHP/ondemand-go-bootcamp/model"
+	"github.com/GerardoHP/ondemand-go-bootcamp/presenter"
+	"github.com/GerardoHP/ondemand-go-bootcamp/repository"
+	"github.com/go-resty/resty/v2"
+)
+
+type Client interface {
+	Get(url string) (*resty.Response, error)
+}
+
+// Implementation of the interface PokemonInteractor
+type pokemonInteractor struct {
+	PokemonRepository repository.Pokemon
+	PokemonPresenter  presenter.Pokemon
+	client            Client
+}
+
+// The interface that holds all the pokemon interactions
+type Pokemon interface {
+	Get(p []*entity.Pokemon) ([]*entity.Pokemon, error)
+	GetPokemon(p string) (*entity.Pokemon, error)
+}
+
+// Returns a new PokemonInteractor instance
+func NewPokemonInteractor(r repository.Pokemon, p presenter.Pokemon, c Client) Pokemon {
+	return &pokemonInteractor{PokemonRepository: r, PokemonPresenter: p, client: c}
+}
+
+// Gets all the pokemons found in the repository
+func (pk *pokemonInteractor) Get(p []*entity.Pokemon) ([]*entity.Pokemon, error) {
+	p, err := pk.PokemonRepository.FindAll(p)
+	if err != nil {
+		return nil, err
+	}
+
+	return pk.PokemonPresenter.ResponsePresenter(p), nil
+}
+
+// Gets a specific pokemon
+func (pk *pokemonInteractor) GetPokemon(p string) (*entity.Pokemon, error) {
+	pokemon, err := pk.PokemonRepository.FindByName(p)
+	if err != nil {
+		return nil, err
+	}
+
+	if pokemon == nil {
+		newPokemon, getPokemonError := getPokemonDetail(p, pk.client)
+		if getPokemonError != nil {
+			return nil, getPokemonError
+		}
+
+		if newPokemon != nil {
+			pokemon, _ = pk.PokemonRepository.Add(newPokemon)
+		}
+
+	}
+
+	return pokemon, nil
+}
+
+// Gets a pokemon detail from a service
+func getPokemonDetail(name string, c Client) (*entity.Pokemon, error) {
+	url := fmt.Sprintf("https://pokeapi.co/api/v2/pokemon/%v", strings.ToLower(name))
+	if c == nil {
+		client := resty.New()
+		c = client.R().EnableTrace()
+	}
+
+	resp, err := c.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.RawResponse.StatusCode == 404 {
+		return nil, errors.New("Not Found")
+	}
+
+	if resp.RawResponse.StatusCode != 200 {
+		return nil, errors.New("Other error")
```

Comment:
To avoid using magic numbers, you can use HTTP.StatusCode

Reply:
Agree